src/sofa/pbrpc/web_service.cc

```diff
@@ -0,0 +1,509 @@
+// Copyright (c) 2015 Baidu.com, Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Author: shichengyi@baidu.com (Shi Chengyi)
+
+#include <sofa/pbrpc/web_service.h>
+#include <sofa/pbrpc/http.h>
+#include <sofa/pbrpc/http_rpc_request.h>
+#include <sofa/pbrpc/rpc_server_impl.h>
+#include <sofa/pbrpc/rpc_server_stream.h>
+
+namespace sofa {
+namespace pbrpc {
+
+static std::string GetHostName(const std::string& ip)
+{
+    struct sockaddr_in addr;
+    struct hostent *host;
+    if (inet_aton(ip.c_str(), &addr.sin_addr) != 0)
+    {
+        host = gethostbyaddr((char*)&addr.sin_addr, 4, AF_INET); 
+    }
+    else
+    {
+        host = gethostbyname(ip.c_str());
+    }
+    if (host != NULL)
+    {
+        return host->h_name;
+    }
+    else
+    {
+        return ip;
+    }
+}
+
+template <class T>
+static std::string format_number(T num)
+{
+    std::ostringstream oss;
+    oss << num;
+    std::string str = oss.str();
+    size_t len = str.size();
+    std::string ret;
+    for (size_t i = 0; i < str.size(); ++i)
+    {
+        if (i != 0 && len % 3 == 0)
+        {
+            ret.push_back(',');
+        }
+        ret.push_back(str.at(i));
+        --len;
+    }
+    return ret;
+}
+
+WebService::WebService(const ServicePoolWPtr& service_pool)
+    : _service_pool(service_pool)
+    , _servlet_map_lock()
+    , _servlet_map(new ServletMap())
+    , _default_home(NULL)
+    , _default_options(NULL)
+    , _default_status(NULL)
+    , _default_services(NULL)
+    , _default_service(NULL)
+{ }
+
+WebService::~WebService()
+{
+    delete _default_home;
+    _default_home = NULL;
+
+    delete _default_options;
+    _default_options = NULL;
+
+    delete _default_status;
+    _default_status = NULL;
+
+    delete _default_services;
+    _default_services = NULL;
+
+    delete _default_service;
+    _default_service = NULL;
+}
+
+void WebService::Init()
+{
+    _default_home = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultHome);
+
+    RegisterServlet("/", _default_home);
+    RegisterServlet("/home", _default_home);
+
+    _default_options = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultOptions);
+    RegisterServlet("/options", _default_options);
+
+    _default_status = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultStatus);
+    RegisterServlet("/status", _default_status);
+
+    _default_services = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultServices);
+    RegisterServlet("/services", _default_services);
+
+    _default_service = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultService);
+    RegisterServlet("/service", _default_service);
+}
+
+void WebService::RegisterServlet(const std::string& path, Servlet servlet)
```

Comment:
RegisterServlet() should return bool: maybe the same path has already been registered